ignite/metrics/ssim.py

```diff
@@ -157,18 +159,34 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
                 f"Expected y_pred and y to have BxCxHxW shape. Got y_pred: {y_pred.shape} and y: {y.shape}."
             )
 
-        channel = y_pred.size(1)
-        if len(self._kernel.shape) < 4:
-            self._kernel = self._kernel.expand(channel, 1, -1, -1).to(device=y_pred.device)
+        if y_pred.device != self._kernel.device:
```

Comment:
Thanks for trying this approach with _kernel and _expanded_kernel !
Few remarks, with now 2 kernel attributes, we can easily get them unsynced. We would like to avoid the situations, so let's code this part as following:
- we should we define _expanded_kernel in the very beginning
```
        nb_channels = y_pred.size(1)
        if self._expanded_kernel is None or self._expanded_kernel.shape[0] != nb_channels:
            self._expanded_kernel = self._kernel.expand(nb_channels, 1, -1, -1)
```

- all device and dtype management should happen on _expanded_kernel only
- we should not use _kernel anywhere except when we define  _expanded_kernel

Nit remarks: I would rename  _kernel -> _kernel_2d and _expanded_kernel -> _kernel